### PR TITLE
Custom evaluator for Open-NMT2 model

### PR DESCRIPTION
**Pull request by mzuevx**
Work in progress, do not merge

### CODE REVIEW

**Comment by mzuevx** [comment]
@eaidova , please review

**Comment by eaidova** [comment]
Jenkins ok to test

**Comment by eaidova** [comment]
@mzuevx please fix linting issues

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+        pass
+
+    def fit_to_input(self, input_data):
+        pass
+
+    def propagate_output(self, output_data):
+        pass
+
+
+class BaseDLSDKModel(BaseModel):
```
Comment (eaidova) [inline tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py:85]: please use BaseDLSDKModel from base_models (at least inherit your class from it)

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+        return raw_outputs, predictions
+
+    def reset(self):
+        self.processing_frames_buffer = []
+        if self._encoder_predictions is not None:
+            self._encoder_predictions = []
+
+    def release(self):
+        self.encoder.release()
+        self.decoder.release()
```
Comment (eaidova) [inline tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py:314]: do you need also release decoder1?

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+        if model.suffix not in accepted_suffixes:
+            raise ConfigError('Models with following suffixes are allowed: {}'.format(accepted_suffixes))
+        print_info('{} - Found model: {}'.format(self.default_model_suffix, model))
+
+        return model
+
+
+class EncoderONNXModel(CommonONNXModel):
+    default_model_suffix = 'encoder'
+    # input_layers = ['src', 'src_len']
```
Comment (eaidova) [inline tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py:654]: please remove commented

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py (hunk 1 above)
Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: Fixed

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py (hunk 2 above)
Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: Decoder1 is debugging artefact, removed.

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+                decoder['_model_is_blob'] = is_blob
+            if not contains_any(generator, ['model', 'onnx_model']) and models_args:
+                generator['model'] = models_args[2 if len(models_args) > 2 else 0]
+                generator['_model_is_blob'] = is_blob
+            network_info.update({'encoder': encoder, 'decoder': decoder, 'generator': generator})
+        if not contains_all(network_info, ['encoder', 'decoder', 'generator']) and not delayed_model_loading:
+            raise ConfigError('network_info should contain encoder, decoder and generator fields')
+        self.encoder = create_model('encoder', network_info['encoder'], launcher, delayed_model_loading)
+        self.decoder = create_model('decoder', network_info['decoder'], launcher, delayed_model_loading)
+        self.decoder1 = create_model('decoder', network_info['decoder'], launcher, delayed_model_loading)
```
[the remark this answers is not in the document]
Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: Decoder1 is debugging artefact, removed

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py (hunk 3 above)
Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: fixed

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        predictions, raw_outputs = [], []
+
+        for data in input_data:
+            decode_strategy = BeamSearch(self.decoder.network_info)
+
+            src_len = np.array([len(data)])
+            h, c, memory = self.encoder.predict(identifiers, {'src': np.array([[[t]] for t in data]),
+                                                              'src_len': src_len})
+            if encoder_callback:
+                encoder_callback((h, c, memory, src_len))
```
Comment (eaidova) [inline tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py:123]: not sure that you use this callback properly
it is callback for providing  opportunity to calculate statistics in POT, it requires to provide full dictionary of outputs, not only meaningful for pipeline (pot can mark additional layers as outputs for getting their calculation results

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    return_layers = []
+
+    def predict(self, identifiers, input_data, callback=None):
+        input_data = self.fit_to_input(input_data)
+        results = self.exec_network.infer(input_data)
+        self.propagate_output(results)
+        names = self.return_layers if len(self.return_layers) > 0 else self.output_layers
+        return tuple([results[name] for name in names])
+
+    def release(self):
```
Comment (eaidova) [inline tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py:212]: why do you need overload release method?

Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: Fixed

File: tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py (hunk 5 above)
Comment (mzuevx) [reply tools/accuracy_checker/openvino/tools/accuracy_checker/evaluators/custom_evaluators/opennmt_encoder_decoder_generator_evaluator.py]: fixed

**Review by eaidova — APPROVED** [review]

**Pull request merged**